File: fittracker-backend/app/models/challenges.py
from datetime import datetime, timezone, date
from typing import Optional
import sqlalchemy as sa
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class Challenge(Base):
    __tablename__ = "challenges"
    id: Mapped[int] = mapped_column(sa.Integer, primary_key=True, index=True)
    creator_id: Mapped[int] = mapped_column(sa.Integer, sa.ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    title: Mapped[str] = mapped_column(sa.String(200), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(sa.Text)
    challenge_type: Mapped[str] = mapped_column(sa.String(30), nullable=False)  # total_workouts / total_volume / streak
    target_value: Mapped[float] = mapped_column(sa.Numeric(10, 2), nullable=False)
    start_date: Mapped[date] = mapped_column(sa.Date, nullable=False)
    end_date: Mapped[date] = mapped_column(sa.Date, nullable=False)
    is_public: Mapped[bool] = mapped_column(sa.Boolean, default=True, server_default="1")
    status: Mapped[str] = mapped_column(sa.String(20), default="active", server_default="active", index=True)  # active/completed/cancelled
    winner_user_id: Mapped[Optional[int]] = mapped_column(sa.Integer, sa.ForeignKey("users.id", ondelete="SET NULL"))
    created_at: Mapped[datetime] = mapped_column(sa.TIMESTAMP(timezone=True), default=lambda: datetime.now(timezone.utc), server_default=sa.func.now())

    participants: Mapped[list["ChallengeParticipant"]] = relationship(
        "ChallengeParticipant", back_populates="challenge", cascade="all, delete-orphan"
    )


class ChallengeParticipant(Base):
    __tablename__ = "challenge_participants"
    id: Mapped[int] = mapped_column(sa.Integer, primary_key=True)
    challenge_id: Mapped[int] = mapped_column(sa.Integer, sa.ForeignKey("challenges.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id: Mapped[int] = mapped_column(sa.Integer, sa.ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    current_progress: Mapped[float] = mapped_column(sa.Numeric(10, 2), default=0, server_default="0")
    rank: Mapped[Optional[int]] = mapped_column(sa.Integer)
    joined_at: Mapped[datetime] = mapped_column(sa.TIMESTAMP(timezone=True), default=lambda: datetime.now(timezone.utc), server_default=sa.func.now())
    __table_args__ = (sa.UniqueConstraint("challenge_id", "user_id", name="uq_challenge_participant"),)

    challenge: Mapped["Challenge"] = relationship("Challenge", back_populates="participants")

File: fittracker-frontend/src/api/challenges.ts
import { apiClient } from './client';

export interface Challenge {
  id: number;
  creator_id: number;
  title: string;
  description: string | null;
  challenge_type: 'total_workouts' | 'total_volume' | 'streak';
  target_value: number;
  start_date: string;
  end_date: string;
  is_public: boolean;
  status: 'active' | 'completed' | 'cancelled';
  winner_user_id: number | null;
  created_at: string;
  participant_count: number;
  is_joined: boolean;
  my_progress: number | null;
  my_rank: number | null;
}

export interface LeaderboardEntry {
  user_id: number;
  username: string;
  avatar_url: string | null;
  current_progress: number;
  rank: number | null;
  joined_at: string;
}

export interface ChallengeCreate {
  title: string;
  description?: string;
  challenge_type: 'total_workouts' | 'total_volume' | 'streak';
  target_value: number;
  start_date: string;
  end_date: string;
  is_public?: boolean;
}

export const challengesApi = {
  list: (params: { status?: string; mine?: boolean } = {}): Promise<Challenge[]> => {
    const q = new URLSearchParams();
    if (params.status) q.set('status', params.status);
    if (params.mine !== undefined) q.set('mine', String(params.mine));
    return apiClient.get<Challenge[]>(`/api/challenges?${q}`).then(r => r.data);
  },
  get: (id: number): Promise<Challenge & { participants: LeaderboardEntry[] }> =>
    apiClient.get(`/api/challenges/${id}`).then(r => r.data),
  create: (data: ChallengeCreate): Promise<Challenge> =>
    apiClient.post<Challenge>('/api/challenges', data).then(r => r.data),
  update: (id: number, data: Partial<ChallengeCreate>): Promise<Challenge> =>
    apiClient.put<Challenge>(`/api/challenges/${id}`, data).then(r => r.data),
  delete: (id: number): Promise<void> =>
    apiClient.delete(`/api/challenges/${id}`).then(() => undefined),
  join: (id: number): Promise<void> =>
    apiClient.post(`/api/challenges/${id}/join`).then(() => undefined),
  leave: (id: number): Promise<void> =>
    apiClient.delete(`/api/challenges/${id}/leave`).then(() => undefined),
  leaderboard: (id: number): Promise<LeaderboardEntry[]> =>
    apiClient.get<LeaderboardEntry[]>(`/api/challenges/${id}/leaderboard`).then(r => r.data),
};
